docs/2024/puzzles/day09.md

```diff
@@ -1,15 +1,320 @@
 import Solver from "../../../../../website/src/components/Solver.js"
 
+by [@dyvrl](https://github.com/dyvrl)
+
 # Day 9: Disk Fragmenter
 
 ## Puzzle description
 
 https://adventofcode.com/2024/day/9
 
+## Solution Summary
+
+1. Convert the input to a disk representation:
+  - `part1`: A sequence of optional file indices
+  - `part2`: A sequence of indivisible file/free-space blocks
+2. Create a compacted representation of this disk: Starting from the end of the disk,
+  - `part1`: Move individual file indices to the leftmost free space
+  - `part2`: Move file blocks to the to the leftmost free block, if any
+3. Compute the checksum of the resulting disk
+
+### Part 1
+
+Each part will define its own `Disk` type. For `part1`, this will simply be a `Seq[Option[Int]]`, where each charater has an assigned value:
+- `Some(index)` for file blocks, with their corresponding index
+- `None` for free blocks
+
+Our main driver converts the input to a `Disk`, create a new compact `Disk` from it and finally computes its checksum
+
+```scala
+def part1(input: String): Long =
+
+  type Disk = Seq[Option[Int]]
+  extension(disk: Disk)
+    def checksum: Long = ???
+
+  def createDisk(input: String): Disk = ???
+
+  def compact(disk: Disk): Disk = ???
+
+  val disk = createDisk(input)
+  compact(disk).checksum
+```
+
+Let's first implement `checksum`: It is the sum of each file ID times its position in the disk
+```scala
+extension(disk: Disk)
+  def checksum: Long = disk
+    .zipWithIndex
+    .map(_.getOrElse(0).toLong * _) // Free blocks are mapped to 0
+    .sum
+```
+
+To create our disk from the input, we need to:
+- Convert our input to a `List[Int]` ranging from 0 to 9
+- Group elements 2 by 2 to create pairs of (file, free) blocks count
+- Zip these groups with their indices
+- Unroll these pairs into an actual sequence with the correct number of elements
+- Concatenate these newly created sequences
+```scala
+def createDisk(input: String): Disk =
+    val intInput = input.toList.map(_ - '0') // Convert characters to int [0, 9]
+    val fileFreeGroups = intInput.grouped(2).toVector // Last group will contain a single element
+    val zippedGroups = fileFreeGroups.zipWithIndex
+    val disk = zippedGroups.flatMap:
+      case (List(fileN, freeN), i) => 
+        // File block followed by free block
+        List.fill(fileN)(Some(i)) ::: List.fill(freeN)(None)
+      case (List(fileN), i) => 
+        // Final file block
+        List.fill(fileN)(Some(i))
+      case _ => Nil
+    return disk
```

Comment:
No need for `return` here, just make `zippedGroups.flatMap` from above the last expression in the method.